File: packages/ops-thread-fsm/lib/ops_thread_fsm/discussion.py
"""Discussion loop checks for ops-thread-fsm.

This module does not run ChatGPT or send messages. It only classifies whether a
discussion-required task has enough same-revision responses to converge and
renders controller-owned facilitation actions.
"""
from __future__ import annotations

import re
from typing import Any

from .state_model import next_action_for, permissions_for

NO_OBJECTIONS = "NO_UNRESOLVED_OBJECTIONS"
UNRESOLVED = "UNRESOLVED_OBJECTIONS"


def _as_list(value: Any) -> list[Any]:
    if value is None:
        return []
    if isinstance(value, list):
        return value
    return [value]


def _token(value: Any) -> str:
    return str(value or "").strip()


def _row(
    classification: str,
    next_state: str,
    *,
    discussion_id: str = "",
    proposal_revision: str = "",
    evidence: list[str] | None = None,
    missing: list[str] | None = None,
    objections: list[dict[str, Any]] | None = None,
    responses: list[dict[str, Any]] | None = None,
    retry: bool = False,
    reason: str | None = None,
    discussion_complete: bool = False,
) -> dict[str, Any]:
    return {
        "kind": "ops-thread-fsm.discussionCheck.v1",
        "classification": classification,
        "stateKind": classification,
        "nextStateKind": next_state,
        "discussionId": discussion_id,
        "proposalRevision": proposal_revision,
        "discussionComplete": discussion_complete,
        "missingCounterparties": missing or [],
        "objections": objections or [],
        "responses": responses or [],
        "evidence": evidence or [],
        "retry": retry,
        "retryReason": reason,
        "permissions": permissions_for(next_state),
        "writes": False,
        "sends": False,
        "nextAction": next_action_for(next_state),
    }


def _verdict(response: dict[str, Any]) -> str:
    for key in ("verdict", "status", "result", "classification"):
        value = _token(response.get(key))
        if value in {NO_OBJECTIONS, UNRESOLVED}:
            return value
    text = _token(response.get("text"))
    first = next((line.strip() for line in text.splitlines() if line.strip()), "")
    if first in {NO_OBJECTIONS, UNRESOLVED}:
        return first
    return ""


def _text_contains_marker(text: str, marker: str) -> bool:
    marker = _token(marker)
    if not marker:
        return False
    if marker in {line.strip() for line in text.splitlines()}:
        return True
    pattern = rf"(?<![A-Za-z0-9_]){re.escape(marker)}(?![A-Za-z0-9_])"
    return re.search(pattern, text, flags=re.IGNORECASE) is not None


def _text_contains_any(text: str, markers: list[str]) -> bool:
    return any(_text_contains_marker(text, marker) for marker in markers)


def _normalize_marker_responses(value: dict[str, Any]) -> list[dict[str, Any]]:
    accepted = [_token(item) for item in _as_list(value.get("acceptedMarkers")) if _token(item)]
    rejected = [_token(item) for item in _as_list(value.get("objectionMarkers")) if _token(item)]
    accepted = accepted or [NO_OBJECTIONS]
    rejected = rejected or [UNRESOLVED]

    responses = []
    for response in _as_list(value.get("responses")):
        if not isinstance(response, dict):
            continue
        item = dict(response)
        if _verdict(item):
            responses.append(item)
            continue
        text = _token(item.get("assistantText") or item.get("text") or item.get("responseText"))
        has_objection = _text_contains_any(text, rejected)
        has_acceptance = _text_contains_any(text, accepted)
        if has_objection:
            item["verdict"] = UNRESOLVED
            item.setdefault("objections", [{"objectionText": "objection marker matched"}])
        elif has_acceptance:
            item["verdict"] = NO_OBJECTIONS
        responses.append(item)
    return responses


def check_discussion_value(value: Any) -> dict[str, Any]:
    if not isinstance(value, dict):
        return _row("insufficient-plan", "discussion-response-required", reason="input must be an object")

    discussion_id = _token(value.get("discussionId"))
    revision = _token(value.get("proposalRevision"))
    required = [_token(item) for item in _as_list(value.get("noObjectionsRequiredFrom")) if _token(item)]
    responses = [item for item in _as_list(value.get("responses")) if isinstance(item, dict)]

    missing_fields = []
    if not discussion_id:
        missing_fields.append("discussionId")
    if not revision:
        missing_fields.append("proposalRevision")
    if not required:
        missing_fields.append("noObjectionsRequiredFrom")
    if missing_fields:
        return _row(
            "insufficient-plan",
            "discussion-response-required",
            discussion_id=discussion_id,
            proposal_revision=revision,
            missing=missing_fields,
            retry=True,
            reason="missing discussion loop fields",
        )

    same_revision_by_actor: dict[str, dict[str, Any]] = {}
    stale_or_wrong_revision = []
    for response in responses:
        actor = _token(response.get("actorId") or response.get("actor") or response.get("thread"))
        response_revision = _token(response.get("proposalRevision"))
        if not actor:
            continue
        if response_revision != revision:
            stale_or_wrong_revision.append(actor)
            continue
        same_revision_by_actor[actor] = response

    missing = [actor for actor in required if actor not in same_revision_by_actor]
    if missing:
        evidence = []
        if stale_or_wrong_revision:
            evidence.append("stale-or-wrong-revision:" + ",".join(sorted(stale_or_wrong_revision)))
        return _row(
            "discussion-response-required",
            "discussion-response-required",
            discussion_id=discussion_id,
            proposal_revision=revision,
            missing=missing,
            responses=list(same_revision_by_actor.values()),
            evidence=evidence,
            retry=True,
            reason="missing same-revision response",
        )

    objections = []
    needs_parent = []
    no_objections = []
    for actor in required:
        response = same_revision_by_actor[actor]
        verdict = _verdict(response)
        if verdict == NO_OBJECTIONS:
            no_objections.append(actor)
            continue
        if verdict == UNRESOLVED:
            entries = _as_list(response.get("objections"))
            if not entries and response.get("objectionText"):
                entries = [response]
            for entry in entries:
                if isinstance(entry, dict):
                    objection = dict(entry)
                else:
                    objection = {"objectionText": str(entry)}
                objection.setdefault("actorId", actor)
                objection.setdefault("proposalRevision", revision)
                objections.append(objection)
                if objection.get("requiresParentDecision") is True or objection.get("classification") == "needs-parent":
                    needs_parent.append(objection)
            if not entries:
                objections.append(
                    {
                        "actorId": actor,
                        "proposalRevision": revision,
                        "objectionText": "unresolved objections declared without structured objection entries",
                    }
                )
            continue
        objections.append(
            {
                "actorId": actor,
                "proposalRevision": revision,
                "objectionText": "missing explicit NO_UNRESOLVED_OBJECTIONS or UNRESOLVED_OBJECTIONS verdict",
            }
        )

    if needs_parent:
        return _row(
            "discussion-blocked-needs-parent",
            "discussion-blocked-needs-parent",
            discussion_id=discussion_id,
            proposal_revision=revision,
            objections=objections,
            responses=list(same_revision_by_actor.values()),
            reason="one or more objections require parent decision",
        )

    if objections:
        return _row(
            "discussion-objections-present",
            "discussion-objections-present",
            discussion_id=discussion_id,
            proposal_revision=revision,
            objections=objections,
            responses=list(same_revision_by_actor.values()),
            retry=True,
            reason="same-revision unresolved objections remain",
        )

    return _row(
        "discussion-no-objections-confirmed",
        "discussion-no-objections-confirmed",
        discussion_id=discussion_id,
        proposal_revision=revision,
        evidence=[f"same-revision-no-objections:{','.join(required)}"],
        responses=list(same_revision_by_actor.values()),
        discussion_complete=True,
    )


def _thread_actor_ids(value: dict[str, Any]) -> list[str]:
    thread_ids = []
    for thread in _as_list(value.get("threads")):
        if isinstance(thread, dict):
            actor_id = _token(thread.get("actorId") or thread.get("actor") or thread.get("thread"))
            if actor_id:
                thread_ids.append(actor_id)
    return thread_ids


def _has_purpose_lineage(value: dict[str, Any]) -> bool:
    lineage = value.get("purposeLineage")
    if not isinstance(lineage, dict):
        return False
    if all(_token(lineage.get(key)) for key in ("purpose", "metaPurpose", "metaMetaPurpose", "metaMetaMetaPurpose")):
        return True
    depths = lineage.get("depths") if isinstance(lineage.get("depths"), dict) else lineage
    return all(_token(depths.get(str(depth)) or depths.get(depth) or depths.get(f"purposeDepth={depth}")) for depth in (3, 2, 1, 0))


def _missing_facilitation_fields(value: dict[str, Any]) -> list[str]:
    missing = []
    for key in ("discussionId", "proposalRevision"):
        if not _token(value.get(key)):
            missing.append(key)
    if not _token(value.get("versionedProposalRef") or value.get("projectSourceEntrypoint")):
        missing.append("versionedProposalRef or projectSourceEntrypoint")
    if not _token(value.get("policySnapshotRef")) and not _as_list(value.get("policyRefs")):
        missing.append("policySnapshotRef or policyRefs")
    if not _has_purpose_lineage(value):
        missing.append("purposeLineage depth 3..0")
    if not _as_list(value.get("reviewQualityChecks")):
        missing.append("reviewQualityChecks")
    if len(_thread_actor_ids(value)) < 2:
        missing.append("threads[2+] with actorId")
    return missing


def _control_for_thread(value: dict[str, Any], thread: dict[str, Any]) -> dict[str, Any]:
    actor_id = _token(thread.get("actorId") or thread.get("actor") or thread.get("thread"))
    thread_function = _token(thread.get("threadFunction") or value.get("threadFunction") or "impl-review")
    return {
        "actorId": actor_id,
        "threadUrl": _token(thread.get("threadUrl") or thread.get("url")),
        "threadFunction": thread_function,
        "sendVia": "ops-cdp-core:project-thread-send",
        "inlinePolicy": "pointer-only",
        "controlText": "\n".join(
            [
                f"role.chatgpt.thread: {thread_function}",
                "",
                "Read the Project Source entrypoint: "
                + _token(value.get("projectSourceEntrypoint") or value.get("versionedProposalRef")),
                "This is direct cross-discussion. Read peer replies directly, not facilitator synthesis.",
                "If objections remain, return UNRESOLVED_OBJECTIONS with structured objections.",
                "If no objections remain for proposalRevision "
                + _token(value.get("proposalRevision"))
                + ", return NO_UNRESOLVED_OBJECTIONS.",
            ]
        ),
    }


def facilitate_discussion_value(value: Any) -> dict[str, Any]:
    if not isinstance(value, dict):
        return {
            "kind": "ops-thread-fsm.discussionFacilitation.v1",
            "classification": "facilitation-context-incomplete",
            "missingFields": ["input object"],
            "writes": False,
            "sends": False,
            "nextAction": "provide a discussion facilitation object",
        }

    missing = _missing_facilitation_fields(value)
    discussion_id = _token(value.get("discussionId"))
    revision = _token(value.get("proposalRevision"))
    required = [_token(item) for item in _as_list(value.get("noObjectionsRequiredFrom")) if _token(item)]
    required = required or _thread_actor_ids(value)

    if missing:
        return {
            "kind": "ops-thread-fsm.discussionFacilitation.v1",
            "classification": "facilitation-context-incomplete",
            "discussionId": discussion_id,
            "proposalRevision": revision,
            "missingFields": missing,
            "writes": False,
            "sends": False,
            "nextAction": "add the missing Project Source bootstrap fields before starting discussion",
        }

    normalized = dict(value)
    normalized["noObjectionsRequiredFrom"] = required
    normalized["responses"] = _normalize_marker_responses(value)
    check = check_discussion_value(normalized)

    controls = [_control_for_thread(value, thread) for thread in _as_list(value.get("threads")) if isinstance(thread, dict)]
    base = {
        "kind": "ops-thread-fsm.discussionFacilitation.v1",
        "discussionId": discussion_id,
        "proposalRevision": revision,
        "requiredCounterparties": required,
        "versionedProposalRef": _token(value.get("versionedProposalRef") or value.get("projectSourceEntrypoint")),
        "policySnapshotRef": _token(value.get("policySnapshotRef")),
        "projectSourceRequired": True,
        "writes": False,
        "sends": False,
        "transportPackage": "ops-cdp-core",
        "fsmPackage": "ops-thread-fsm",
    }

    if check["classification"] == "discussion-no-objections-confirmed":
        return {
            **base,
            "classification": "facilitation-no-objections-confirmed",
            "discussionComplete": True,
            "acceptedProposalRef": _token(value.get("versionedProposalRef") or value.get("projectSourceEntrypoint")),
            "evidence": check["evidence"],
            "responses": check["responses"],
            "nextAction": "import this verdict as review evidence; this is not localize, merge, push, or cleanup approval",
        }

    if check["classification"] == "discussion-objections-present":
        return {
            **base,
            "classification": "facilitation-revision-update-required",
            "discussionComplete": False,
            "objections": check["objections"],
            "responses": check["responses"],
            "requiredNextArtifact": "new versioned proposal with accepted/rejected/modified objection handling",
            "nextAction": "create a new proposalRevision in Project Source and run another direct cross-discussion round",
        }

    if check["classification"] == "discussion-blocked-needs-parent":
        return {
            **base,
            "classification": "facilitation-parent-decision-required",
            "discussionComplete": False,
            "objections": check["objections"],
            "responses": check["responses"],
            "nextAction": "ask parentActor to decide the needs-parent objection before continuing",
        }

    return {
        **base,
        "classification": "facilitation-round-send-required",
        "discussionComplete": False,
        "missingCounterparties": check["missingCounterparties"],
        "threadControls": controls,
        "readbackPolicy": "wait >=300s before semantic readback",
        "nextAction": "upload/update Project Source if needed, send pointer-only controls, wait >=300s, then import assistant replies and re-run",
    }


__all__ = ["NO_OBJECTIONS", "UNRESOLVED", "check_discussion_value", "facilitate_discussion_value"]

File: packages/ops-thread-fsm/lib/discussion.mjs
// Discussion loop checks for ops-thread-fsm.
//
// This module does not run ChatGPT or send messages. It only classifies whether a
// discussion-required task has enough same-revision responses to converge and
// renders controller-owned facilitation actions.
//
// Node ESM port of discussion.py (stdlib only, behavior-identical).

import { nextActionFor, permissionsFor } from "./state_model.mjs";
import { pyOr, splitlines } from "./pyhelpers.mjs";

export const NO_OBJECTIONS = "NO_UNRESOLVED_OBJECTIONS";
export const UNRESOLVED = "UNRESOLVED_OBJECTIONS";

function isPlainObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function get(obj, key) {
  if (isPlainObject(obj) && Object.prototype.hasOwnProperty.call(obj, key)) {
    return obj[key];
  }
  return undefined;
}

function pyStr(value) {
  if (typeof value === "string") return value;
  if (value === true) return "True";
  if (value === false) return "False";
  if (value === null || value === undefined) return "None";
  if (typeof value === "number") return String(value);
  return JSON.stringify(value);
}

function _asList(value) {
  if (value === null || value === undefined) {
    return [];
  }
  if (Array.isArray(value)) {
    return value;
  }
  return [value];
}

function _token(value) {
  return pyStr(pyOr(value, "")).trim();
}

function _row(
  classification,
  nextState,
  {
    discussionId = "",
    proposalRevision = "",
    evidence = null,
    missing = null,
    objections = null,
    responses = null,
    retry = false,
    reason = null,
    discussionComplete = false,
  } = {},
) {
  return {
    kind: "ops-thread-fsm.discussionCheck.v1",
    classification,
    stateKind: classification,
    nextStateKind: nextState,
    discussionId,
    proposalRevision,
    discussionComplete,
    missingCounterparties: pyOr(missing, []),
    objections: pyOr(objections, []),
    responses: pyOr(responses, []),
    evidence: pyOr(evidence, []),
    retry,
    retryReason: reason,
    permissions: permissionsFor(nextState),
    writes: false,
    sends: false,
    nextAction: nextActionFor(nextState),
  };
}

function _verdict(response) {
  for (const key of ["verdict", "status", "result", "classification"]) {
    const value = _token(get(response, key));
    if (value === NO_OBJECTIONS || value === UNRESOLVED) {
      return value;
    }
  }
  const text = _token(get(response, "text"));
  let first = "";
  for (const line of splitlines(text)) {
    if (line.trim()) {
      first = line.trim();
      break;
    }
  }
  if (first === NO_OBJECTIONS || first === UNRESOLVED) {
    return first;
  }
  return "";
}

function _textContainsAny(text, markers) {
  const haystack = text.toLowerCase();
  return markers.some((marker) => marker && haystack.includes(marker.toLowerCase()));
}

function _normalizeMarkerResponses(value) {
  let accepted = _asList(get(value, "acceptedMarkers")).map((item) => _token(item)).filter((item) => item);
  let rejected = _asList(get(value, "objectionMarkers")).map((item) => _token(item)).filter((item) => item);
  accepted = pyOr(accepted, [NO_OBJECTIONS]);
  rejected = pyOr(rejected, [UNRESOLVED]);

  const responses = [];
  for (const response of _asList(get(value, "responses"))) {
    if (!isPlainObject(response)) {
      continue;
    }
    const item = { ...response };
    if (_verdict(item)) {
      responses.push(item);
      continue;
    }
    const text = _token(
      pyOr(get(item, "assistantText"), pyOr(get(item, "text"), get(item, "responseText"))),
    );
    const hasObjection = _textContainsAny(text, rejected);
    const hasAcceptance = _textContainsAny(text, accepted);
    if (hasObjection) {
      item.verdict = UNRESOLVED;
      if (!Object.prototype.hasOwnProperty.call(item, "objections")) {
        item.objections = [{ objectionText: "objection marker matched" }];
      }
    } else if (hasAcceptance) {
      item.verdict = NO_OBJECTIONS;
    }
    responses.push(item);
  }
  return responses;
}

export function checkDiscussionValue(value) {
  if (!isPlainObject(value)) {
    return _row("insufficient-plan", "discussion-response-required", { reason: "input must be an object" });
  }

  const discussionId = _token(get(value, "discussionId"));
  const revision = _token(get(value, "proposalRevision"));
  const required = _asList(get(value, "noObjectionsRequiredFrom"))
    .map((item) => _token(item))
    .filter((item) => item);
  const responses = _asList(get(value, "responses")).filter((item) => isPlainObject(item));

  const missingFields = [];
  if (!discussionId) {
    missingFields.push("discussionId");
  }
  if (!revision) {
    missingFields.push("proposalRevision");
  }
  if (!required.length) {
    missingFields.push("noObjectionsRequiredFrom");
  }
  if (missingFields.length) {
    return _row("insufficient-plan", "discussion-response-required", {
      discussionId,
      proposalRevision: revision,
      missing: missingFields,
      retry: true,
      reason: "missing discussion loop fields",
    });
  }

  const sameRevisionByActor = new Map();
  const staleOrWrongRevision = [];
  for (const response of responses) {
    const actor = _token(
      pyOr(get(response, "actorId"), pyOr(get(response, "actor"), get(response, "thread"))),
    );
    const responseRevision = _token(get(response, "proposalRevision"));
    if (!actor) {
      continue;
    }
    if (responseRevision !== revision) {
      staleOrWrongRevision.push(actor);
      continue;
    }
    sameRevisionByActor.set(actor, response);
  }

  const missing = required.filter((actor) => !sameRevisionByActor.has(actor));
  if (missing.length) {
    const evidence = [];
    if (staleOrWrongRevision.length) {
      evidence.push("stale-or-wrong-revision:" + [...staleOrWrongRevision].sort().join(","));
    }
    return _row("discussion-response-required", "discussion-response-required", {
      discussionId,
      proposalRevision: revision,
      missing,
      responses: [...sameRevisionByActor.values()],
      evidence,
      retry: true,
      reason: "missing same-revision response",
    });
  }

  const objections = [];
  const needsParent = [];
  const noObjections = [];
  for (const actor of required) {
    const response = sameRevisionByActor.get(actor);
    const verdict = _verdict(response);
    if (verdict === NO_OBJECTIONS) {
      noObjections.push(actor);
      continue;
    }
    if (verdict === UNRESOLVED) {
      let entries = _asList(get(response, "objections"));
      if (!entries.length && get(response, "objectionText")) {
        entries = [response];
      }
      for (const entry of entries) {
        let objection;
        if (isPlainObject(entry)) {
          objection = { ...entry };
        } else {
          objection = { objectionText: pyStr(entry) };
        }
        if (!Object.prototype.hasOwnProperty.call(objection, "actorId")) {
          objection.actorId = actor;
        }
        if (!Object.prototype.hasOwnProperty.call(objection, "proposalRevision")) {
          objection.proposalRevision = revision;
        }
        objections.push(objection);
        if (
          get(objection, "requiresParentDecision") === true ||
          get(objection, "classification") === "needs-parent"
        ) {
          needsParent.push(objection);
        }
      }
      if (!entries.length) {
        objections.push({
          actorId: actor,
          proposalRevision: revision,
          objectionText: "unresolved objections declared without structured objection entries",
        });
      }
      continue;
    }
    objections.push({
      actorId: actor,
      proposalRevision: revision,
      objectionText: "missing explicit NO_UNRESOLVED_OBJECTIONS or UNRESOLVED_OBJECTIONS verdict",
    });
  }

  if (needsParent.length) {
    return _row("discussion-blocked-needs-parent", "discussion-blocked-needs-parent", {
      discussionId,
      proposalRevision: revision,
      objections,
      responses: [...sameRevisionByActor.values()],
      reason: "one or more objections require parent decision",
    });
  }

  if (objections.length) {
    return _row("discussion-objections-present", "discussion-objections-present", {
      discussionId,
      proposalRevision: revision,
      objections,
      responses: [...sameRevisionByActor.values()],
      retry: true,
      reason: "same-revision unresolved objections remain",
    });
  }

  return _row("discussion-no-objections-confirmed", "discussion-no-objections-confirmed", {
    discussionId,
    proposalRevision: revision,
    evidence: [`same-revision-no-objections:${required.join(",")}`],
    responses: [...sameRevisionByActor.values()],
    discussionComplete: true,
  });
}

function _threadActorIds(value) {
  const threadIds = [];
  for (const thread of _asList(get(value, "threads"))) {
    if (isPlainObject(thread)) {
      const actorId = _token(
        pyOr(get(thread, "actorId"), pyOr(get(thread, "actor"), get(thread, "thread"))),
      );
      if (actorId) {
        threadIds.push(actorId);
      }
    }
  }
  return threadIds;
}

function _hasPurposeLineage(value) {
  const lineage = get(value, "purposeLineage");
  if (!isPlainObject(lineage)) {
    return false;
  }
  if (["purpose", "metaPurpose", "metaMetaPurpose", "metaMetaMetaPurpose"].every((key) => _token(get(lineage, key)))) {
    return true;
  }
  const depthsRaw = get(lineage, "depths");
  const depths = isPlainObject(depthsRaw) ? depthsRaw : lineage;
  return [3, 2, 1, 0].every((depth) =>
    _token(pyOr(get(depths, String(depth)), pyOr(get(depths, depth), get(depths, `purposeDepth=${depth}`)))),
  );
}

function _missingFacilitationFields(value) {
  const missing = [];
  for (const key of ["discussionId", "proposalRevision"]) {
    if (!_token(get(value, key))) {
      missing.push(key);
    }
  }
  if (!_token(pyOr(get(value, "versionedProposalRef"), get(value, "projectSourceEntrypoint")))) {
    missing.push("versionedProposalRef or projectSourceEntrypoint");
  }
  if (!_token(get(value, "policySnapshotRef")) && !_asList(get(value, "policyRefs")).length) {
    missing.push("policySnapshotRef or policyRefs");
  }
  if (!_hasPurposeLineage(value)) {
    missing.push("purposeLineage depth 3..0");
  }
  if (!_asList(get(value, "reviewQualityChecks")).length) {
    missing.push("reviewQualityChecks");
  }
  if (_threadActorIds(value).length < 2) {
    missing.push("threads[2+] with actorId");
  }
  return missing;
}

function _controlForThread(value, thread) {
  const actorId = _token(pyOr(get(thread, "actorId"), pyOr(get(thread, "actor"), get(thread, "thread"))));
  const threadFunction = _token(
    pyOr(get(thread, "threadFunction"), pyOr(get(value, "threadFunction"), "impl-review")),
  );
  return {
    actorId,
    threadUrl: _token(pyOr(get(thread, "threadUrl"), get(thread, "url"))),
    threadFunction,
    sendVia: "ops-cdp-core:project-thread-send",
    inlinePolicy: "pointer-only",
    controlText: [
      `role.chatgpt.thread: ${threadFunction}`,
      "",
      "Read the Project Source entrypoint: " +
        _token(pyOr(get(value, "projectSourceEntrypoint"), get(value, "versionedProposalRef"))),
      "This is direct cross-discussion. Read peer replies directly, not facilitator synthesis.",
      "If objections remain, return UNRESOLVED_OBJECTIONS with structured objections.",
      "If no objections remain for proposalRevision " +
        _token(get(value, "proposalRevision")) +
        ", return NO_UNRESOLVED_OBJECTIONS.",
    ].join("\n"),
  };
}

export function facilitateDiscussionValue(value) {
  if (!isPlainObject(value)) {
    return {
      kind: "ops-thread-fsm.discussionFacilitation.v1",
      classification: "facilitation-context-incomplete",
      missingFields: ["input object"],
      writes: false,
      sends: false,
      nextAction: "provide a discussion facilitation object",
    };
  }

  const missing = _missingFacilitationFields(value);
  const discussionId = _token(get(value, "discussionId"));
  const revision = _token(get(value, "proposalRevision"));
  let required = _asList(get(value, "noObjectionsRequiredFrom"))
    .map((item) => _token(item))
    .filter((item) => item);
  required = pyOr(required, _threadActorIds(value));

  if (missing.length) {
    return {
      kind: "ops-thread-fsm.discussionFacilitation.v1",
      classification: "facilitation-context-incomplete",
      discussionId,
      proposalRevision: revision,
      missingFields: missing,
      writes: false,
      sends: false,
      nextAction: "add the missing Project Source bootstrap fields before starting discussion",
    };
  }

  const normalized = { ...value };
  normalized.noObjectionsRequiredFrom = required;
  normalized.responses = _normalizeMarkerResponses(value);
  const check = checkDiscussionValue(normalized);

  const controls = _asList(get(value, "threads"))
    .filter((thread) => isPlainObject(thread))
    .map((thread) => _controlForThread(value, thread));
  const base = {
    kind: "ops-thread-fsm.discussionFacilitation.v1",
    discussionId,
    proposalRevision: revision,
    requiredCounterparties: required,
    versionedProposalRef: _token(pyOr(get(value, "versionedProposalRef"), get(value, "projectSourceEntrypoint"))),
    policySnapshotRef: _token(get(value, "policySnapshotRef")),
    projectSourceRequired: true,
    writes: false,
    sends: false,
    transportPackage: "ops-cdp-core",
    fsmPackage: "ops-thread-fsm",
  };

  if (check.classification === "discussion-no-objections-confirmed") {
    return {
      ...base,
      classification: "facilitation-no-objections-confirmed",
      discussionComplete: true,
      acceptedProposalRef: _token(
        pyOr(get(value, "versionedProposalRef"), get(value, "projectSourceEntrypoint")),
      ),
      evidence: check.evidence,
      responses: check.responses,
      nextAction:
        "import this verdict as review evidence; this is not localize, merge, push, or cleanup approval",
    };
  }

  if (check.classification === "discussion-objections-present") {
    return {
      ...base,
      classification: "facilitation-revision-update-required",
      discussionComplete: false,
      objections: check.objections,
      responses: check.responses,
      requiredNextArtifact: "new versioned proposal with accepted/rejected/modified objection handling",
      nextAction:
        "create a new proposalRevision in Project Source and run another direct cross-discussion round",
    };
  }

  if (check.classification === "discussion-blocked-needs-parent") {
    return {
      ...base,
      classification: "facilitation-parent-decision-required",
      discussionComplete: false,
      objections: check.objections,
      responses: check.responses,
      nextAction: "ask parentActor to decide the needs-parent objection before continuing",
    };
  }

  return {
    ...base,
    classification: "facilitation-round-send-required",
    discussionComplete: false,
    missingCounterparties: check.missingCounterparties,
    threadControls: controls,
    readbackPolicy: "wait >=300s before semantic readback",
    nextAction:
      "upload/update Project Source if needed, send pointer-only controls, wait >=300s, then import assistant replies and re-run",
  };
}
